src/flow/abstractObjectFactories.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { Realm } from "../realm.js";
+import buildExpressionTemplate from "../utils/builder.js";
+import { ObjectCreate, ArrayCreate } from "../methods/index.js";
+import { ValuesDomain } from "../domains/index.js";
+import { Value, AbstractValue, ObjectValue, ArrayValue } from "../values/index.js";
+import invariant from "../invariant.js";
+import { type ObjectTypeTemplate } from "./utils.js";
+
+export function createObject(realm: Realm, shape: ObjectTypeTemplate | null, name: string | null): ObjectValue {
+  let obj = ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+  if (shape != null) {
+    // to get around Flow complaining that shape could be null
+    let shapeThatIsNotNull = shape;
+    Object.keys(shape).forEach((id: string) => {
+      let value = shapeThatIsNotNull[id];
+      invariant(value instanceof Value, "creation of object failed due to object containing non-value properties");
+      obj.$Set(id, value, obj);
+      if (name !== null) {
+        value.intrinsicName = `${name}.${id}`;
+      }
+    });
+  }
+  if (name !== null) {
+    obj.intrinsicName = name;
+  }
+  return obj;
+}
+
+export function createArray(realm: Realm, name: string | null): ArrayValue {
+  let obj = ArrayCreate(realm, 0, realm.intrinsics.ArrayPrototype);
+  if (name !== null) {
+    obj.intrinsicName = name;
+  }
+  return ((obj: any): ArrayValue);
+}
+
+function _createAbstractArray(realm: Realm, name: string | null): AbstractValue {
+  if (name === null) {
+    name = "unknown";
```

Comment:
And then buildExpressionTemplate turns that into a code snippet `unknown`? How can that ever be okay?